torch/_dynamo/output_graph.py

```diff
@@ -2229,6 +2199,296 @@ def maybe_lift_tracked_freevar_to_input(self, arg):
+            self._lift_symbols_in_symint(
+                s,
+                (
+                    TensorPropertySource(src, TensorProperty.SIZE, i)
+                    if src is not None
+                    else None
+                ),
+                before=True,
+            )
+        if t.layout is torch.strided:
+            for i, s in enumerate(t.stride()):
+                self._lift_symbols_in_symint(
+                    s,
+                    (
+                        TensorPropertySource(src, TensorProperty.STRIDE, i)
+                        if src is not None
+                        else None
+                    ),
+                    before=True,
+                )
+            self._lift_symbols_in_symint(
+                t.storage_offset(),
+                (
+                    TensorPropertySource(src, TensorProperty.STORAGE_OFFSET)
+                    if src is not None
+                    else None
+                ),
+                before=True,
+            )
+        elif t.layout is torch.sparse_coo:
+            self._lift_symbols_in_tensor(t._indices(), src)
+            self._lift_symbols_in_tensor(t._values(), src)
+        elif t.layout in {torch.sparse_csr, torch.sparse_bsr}:
+            self._lift_symbols_in_tensor(t.crow_indices(), src)
+            self._lift_symbols_in_tensor(t.col_indices(), src)
+        elif t.layout in {torch.sparse_csc, torch.sparse_bsc}:
+            self._lift_symbols_in_tensor(t.ccol_indices(), src)
+            self._lift_symbols_in_tensor(t.row_indices(), src)
+        if is_traceable_wrapper_subclass(t):
+            attrs, ctx = t.__tensor_flatten__()
+            for attr in attrs:
+                inner_t = getattr(t, attr)
+                self._lift_symbols_in_tensor(
+                    inner_t, AttrSource(src, attr) if src is not None else None
+                )
+
+    # Lookup the proxy in current tracer for each symbol in expressions of s,
+    # See Note [Auto lift basic free symbols when create_graph_input]
+    def lookup_unbound_symbols(self, s: torch.SymInt) -> Set[sympy.Symbol]:
+        free_symbols = s.node.expr.free_symbols
+        if len(free_symbols) == 0:
+            return set()
+
+        to_be_bound = set()
+        for s0 in free_symbols:
+            if s0 not in self.bound_symbols:
+                to_be_bound.add(s0)
+                continue
+
+            proxy = self.bound_symbols[s0]
+            proxy = proxy() if isinstance(proxy, SubgraphTracer.LazyProxy) else proxy
+            assert (
+                proxy.tracer is self
+            ), f"The proxy of symbol {s0} doesn't belong to current tracer."
+            self.bound_symbols[s0] = proxy
+        return to_be_bound
+
+    # See NOTE: [Auto lift basic free symbols when create_graph_input]
+    def _lift_symbols_in_symint(
+        self,
+        s: Union[int, torch.SymInt],
+        source: Optional[Source],
+        before: bool = False,
```

Comment:
What is `before`? It's annoying that this is only used in one of the two branches below. Can we use it in both branches and document this?

Reply:
Sure. "Before" is for inserting the symints before tensors. This ordering is useful in insert_deferred_runtime_asserts, which goes through the placeholders sequentially, and if a symbol in a tensor is not tracked, it will create a sym_size node in the graph for that symint https://github.com/pytorch/pytorch/blob/main/torch/fx/passes/runtime_assert.py#L292. To avoid creating redundant proxies, we need these symints to go before the tensor.